### PR TITLE
Added CUB2011 dataset

### PR DESCRIPTION
**Pull request by vadimkantorov**
For testing some metric learning methods I created a wrapper for CUB2011 dataset (inheriting from CIFAR10 for download functionality and from ImageFolder for directory parsing).

The dataset is small, so it'd be possible to serve it from memory, but haven't needed it yet. Also not sure about the good practice on integrity hashes, just put there whole archive hash and a few image hashes.

Let me know what you think!

### CODE REVIEW

File: torchvision/datasets/cub2011.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import os
+import torch
+import torch.utils.data as data
+import torchvision
+from torchvision.datasets import ImageFolder
+from torchvision.datasets import CIFAR10
+
+class Cub2011(ImageFolder, CIFAR10):
```
Comment (szagoruyko) [inline torchvision/datasets/cub2011.py]: should be all uppercase I think `CUB2011`

**Comment by szagoruyko** [comment]
can you add annotations?

**Comment by vadimkantorov** [comment]
I'll rename the class and will fix lint errors.

About the annotations, every image [seems](http://www.vision.caltech.edu/visipedia/CUB-200-2011.html) to have "15 Part Locations, 312 Binary Attributes, 1 Bounding Box". 
Should we always return all of those?
Do we have an accepted format for bounding boxes? (x1-y1-x2-y2?)

I will check if an image always have 15 part locations, that would simplify the encoding.

**Comment by vadimkantorov** [comment]
VOC pull request indeed uses "[xmin, ymin, xmax, ymax, ind]" format so I'll stick to it.

File: torchvision/datasets/cub2011.py (hunk 1 above)
Comment (alykhantejani) [reply torchvision/datasets/cub2011.py]: Not sure you can leverage the `CIFAR10` dataset here as it assumes 60k images internally with a 50k/10k split. However, CUB2011 only has 11,788 images.

File: torchvision/datasets/cub2011.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    base_folder = 'CUB_200_2011/images'
+    url = 'http://www.vision.caltech.edu/visipedia-data/CUB-200-2011/CUB_200_2011.tgz'
+    filename = 'CUB_200_2011.tgz'
+    tgz_md5 = '97eceeb196236b17998738112f37df78'
+    
+    train_list = [
+        ['001.Black_footed_Albatross/Black_Footed_Albatross_0001_796111.jpg', '4c84da568f89519f84640c54b7fba7c2'],
```
Comment (alykhantejani) [inline torchvision/datasets/cub2011.py]: I don't think we need to hash each image or each directory, the hash over the tar file itself should suffice.

File: torchvision/datasets/cub2011.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        ['002.Laysan_Albatross/Laysan_Albatross_0001_545.jpg', 'e7db63424d0e384dba02aacaf298cdc0'],
+    ]
+    test_list = [
```
Comment (alykhantejani) [inline torchvision/datasets/cub2011.py]: If we plan to levearge `CIFAR10` as a super class (see comment above), then perhaps these can be auto generated from the `train_test_split.txt` in the tarfile.

**Comment by varunagrawal** [comment]
Is this PR still relevant? @vadimkantorov seems to have deleted his torchvision fork.

**Comment by vadimkantorov** [comment]
I am not working on this. Please go ahead if you wish to pick this up (my original CUB2011 dataset code is here: https://github.com/vadimkantorov/metriclearningbench/blob/master/cub2011.py )

Some things you may want to look at:
1. Refactor super-hacky `class CUB2011(ImageFolder, CIFAR10)` way to reuse dataset download steps. Maybe now there's a better API for that.
2. Actually implement "15 Part Locations, 312 Binary Attributes, 1 Bounding Box" loading (but maybe the CUB2011MetricLearning would suffice?)

**Comment by pmeier** [comment]
@vadimkantorov 

> I am not working on this.

Closing this in favor of #4126.
